gmail.com: [iPad] Tapping the file icon when composing an email brings up the keyboard

jesxilin opened the pull request:
#### 283687c657f3eee714f2918de62129ee15bda3b6
<pre>
gmail.com: [iPad] Tapping the file icon when composing an email brings up the keyboard
<a href="https://bugs.webkit.org/show_bug.cgi?id=281194">https://bugs.webkit.org/show_bug.cgi?id=281194</a>
<a href="https://rdar.apple.com/89692326">rdar://89692326</a>

Reviewed by Wenson Hsieh.

Changed `WKContentViewInteraction.mm` to check for repositioning context menu
when keyboard is dismissed. If keyboard is dismissed, we should reposition back
to the bottom where it would be without a keyboard shown.
Fixed positioning of context menu when file icon is clicked in gmail.com [iPad]
so when keyboard is up, we don&apos;t cover the context menu.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _keyboardDidShow]):
(-[WKContentView _keyboardDidRequestDismissal:]):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
(-[WKFileUploadPanel presentWithParameters:resultListener:]):

Canonical link: <a href="https://commits.webkit.org/285153@main">https://commits.webkit.org/285153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/863ec96ce1ca10876f549a860ac7196c0b4bae69

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71671 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51084 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75785 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22876 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58885 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22696 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56607 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15087 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74737 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46347 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61740 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37056 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43008 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21217 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64912 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19570 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77505 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15905 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18751 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64319 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15949 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61773 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64319 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15847 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12473 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6112 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46884 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1663 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47955 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49239 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47697 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -496,7 +498,13 @@ - (void)presentWithParameters:(API::OpenPanelParameters*)parameters resultListen
+
+    auto *webView = [_view webView];
+    if (webView)
+        _interactionPointInWindow = [_view convertPoint:_interactionPoint toView:webView.window];
```
Nit - I think it's actually safe to skip the null check here (since `nil` will default to the content view's window anyways)
```suggestion
    _interactionPointInWindow = [_view convertPoint:_interactionPoint toView:[webView window]];
```

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -4750,6 +4750,9 @@ - (void)_keyboardDidRequestDismissal:(NSNotification *)notification
 {
     if (_isEditable && [self isFirstResponder])
         _keyboardDidRequestDismissal = YES;
+#if USE(UICONTEXTMENU)
```
Nit: Add a newline before this.

whsieh reviewed (approved):
r=mews

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -380,6 +380,8 @@ @implementation WKFileUploadPanel {
     RetainPtr<NSSet<NSString *>> _acceptedUTIs;
     OptionSet<WKFileUploadPanelImagePickerType> _allowedImagePickerTypes;
     CGPoint _interactionPoint;
+    CGPoint _interactionPointInWindow;
```
I don't think this instance variable is necessary? Why not just compute it where it's used?

jesxilin replied:
That's how it was before. It's not computed where it's used anymore b/c of _timing issue_. If I'm correct, the menu would be/shown at the bottom of the web view. The view is then pushed up by the keyboard and the menu position in relation to the _window_ would now be right above the keyboard (even though it's not shown there). Our checks will then say that nothing is obscuring the menu b/c the position is _above_ the keyboard.

pxlcoder replied:
Sorry, I got confused after seeing `webView` variable 😅

Makes sense since the content view is resized!

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -800,15 +808,17 @@ - (void)repositionContextMenuIfNeeded
+    if (!_isMenuPreviouslyRepositioned && !CGRectContainsPoint(inputViewBoundsInWindow, _interactionPointInWindow))
         return;
 
+    _isMenuPreviouslyRepositioned = !_isMenuPreviouslyRepositioned;
```
I don't fully understand why this is being toggled, and not set to a fixed value.

jesxilin replied:
It's used when keyboard dismisses and the menu is/was up. The value flips with each function call. So when keyboard shows, it is True at first. When we dismiss it, the menu is still up and has been previously positioned. We only want those early returns if the menu is coming up for the _first time_ or _if it is not up rn_. Otherwise, early returns will trigger and never fix the position to go back down when keyboard is down.

pxlcoder replied:
What if you instead passed in a boolean parameter to this method that encoded whether the keyboard was being presented or dismissed?

jesxilin replied:
I assume you mean smth like a `isKeyboardBeingDismissed` and we pass in the T/F when we call it in `_keyboardDidRequestDismissal` and  `_keyboardDidShow`. I'll try it!

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -4750,6 +4750,10 @@ - (void)_keyboardDidRequestDismissal:(NSNotification *)notification
         _keyboardDidRequestDismissal = YES;
+
+#if USE(UICONTEXTMENU)
+    [_fileUploadPanel repositionContextMenuIfNeeded:YES];
```
Nit - for a bit more clarity, you could do one of the following:

1. Make the `BOOL` argument an `enum class` instead (e.g. `enum class KeyboardIsDismissing : bool { No, Yes };`), or

2. Rename the method to something along the lines of `- (void)repositionContextMenuIfNeededAfterKeyboardDismissal:(BOOL)afterKeyboardDismissal;`

jesxilin replied:
I can do #2 but what is the idea behind #1 ? Why would it give more clarity?

pxlcoder replied:
Correct, otherwise it's unclear what the boolean actually represents.

For example, it would read like `repositionContextMenuIfNeededAfterKeyboardDismissal:NO`, which is confusing because that call could still reposition the menu.

whsieh replied:
Right — the root problem is that from just looking at the call site (`[_fileUploadPanel repositionContextMenuIfNeeded:YES];`), it's not obvious what `YES` represents.

(1) and (2) both help to mitigate this by adding more context to clarify that `YES` in this case refers to the fact that it denotes whether the call is after keyboard dismissal.

jesxilin replied:
Since repositioning the menu is needed before keyboard dismissal, I'm thinking that the function name shouldn't change now... but (1) definitely makes sense. Smth like `[_fileUploadPanel repositionContextMenuIfNeeded:KeyboardIsDismissing::YES]` which would signify reposition if needed AND given keyboard is dismissing rn?

pxlcoder replied:
Yeah, that seems fine.

jesxilin committed:
gmail.com: [iPad] Tapping the file icon when composing an email brings up the keyboard
https://bugs.webkit.org/show_bug.cgi?id=281194
rdar://89692326

Reviewed by Wenson Hsieh.

Changed `WKContentViewInteraction.mm` to check for repositioning context menu
when keyboard is dismissed. If keyboard is dismissed, we should reposition back
to the bottom where it would be without a keyboard shown.
Fixed positioning of context menu when file icon is clicked in gmail.com [iPad]
so when keyboard is up, we don't cover the context menu.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _keyboardDidShow]):
(-[WKContentView _keyboardDidRequestDismissal:]):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
(-[WKFileUploadPanel presentWithParameters:resultListener:]):

Canonical link: https://commits.webkit.org/285153@main
